connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java

```diff
@@ -0,0 +1,188 @@
+    private final static int BQ_NUMERIC_PRECISION = 38;
+    private final static int BQ_NUMERIC_SCALE = 9;
+    private final static DecimalType NUMERIC_SPARK_TYPE = DataTypes.createDecimalType(
+    BQ_NUMERIC_PRECISION, BQ_NUMERIC_SCALE);
+
+    /** Convert a BigQuery schema to a Spark schema */
+    public static StructType toSpark(Schema schema) {
+        List<StructField> fieldList = schema.getFields().stream()
+                .map(SchemaConverters::convert).collect(Collectors.toList());
+        StructType structType = new StructType(fieldList.toArray(new StructField[0]));
+
+        return structType;
+    }
+
+    public static InternalRow createRowConverter(Schema schema, List<String> namesInOrder, GenericRecord record) {
+        return convertAll(schema.getFields(), record, namesInOrder);
+    }
+
+    static Object convert(Field field, Object value) {
+        if (value == null) {
+            return null;
+        }
+
+        if (field.getMode() == Field.Mode.REPEATED) {
+            // rather than recurring down we strip off the repeated mode
+            // Due to serialization issues, reconstruct the type using reflection:
+            // See: https://github.com/googleapis/google-cloud-java/issues/3942
+            LegacySQLTypeName fType = LegacySQLTypeName.valueOfStrict(field.getType().name());
+            Field nestedField = Field.newBuilder(field.getName(), fType, field.getSubFields())
+                    // As long as this is not repeated it works, but technically arrays cannot contain
+                    // nulls, so select required instead of nullable.
+                    .setMode(Field.Mode.REQUIRED)
+                    .build();
+
+            List<Object> valueList = (List<Object>) value;
+
+            return new GenericArrayData(valueList.stream().map(v -> convert(nestedField, v)).collect(Collectors.toList()));
+        }
+
+        if (LegacySQLTypeName.INTEGER.equals(field.getType()) ||
+                LegacySQLTypeName.FLOAT.equals(field.getType()) ||
+                LegacySQLTypeName.BOOLEAN.equals(field.getType()) ||
+                LegacySQLTypeName.DATE.equals(field.getType()) ||
+                LegacySQLTypeName.TIME.equals(field.getType()) ||
+                LegacySQLTypeName.TIMESTAMP.equals(field.getType())) {
+            return value;
+        }
+
+        if (LegacySQLTypeName.STRING.equals(field.getType()) ||
+                LegacySQLTypeName.DATETIME.equals(field.getType()) ||
+                LegacySQLTypeName.GEOGRAPHY.equals(field.getType())) {
+            return UTF8String.fromBytes(((Utf8)value).getBytes());
+        }
+
+        if (LegacySQLTypeName.BYTES.equals(field.getType())) {
+            return getBytes((ByteBuffer)value);
+        }
+
+        if (LegacySQLTypeName.NUMERIC.equals(field.getType())) {
+            byte[] bytes = getBytes((ByteBuffer)value);
+            BigDecimal b = new BigDecimal(new BigInteger(bytes), BQ_NUMERIC_SCALE);
```

Comment:
@davidrabinowitz is there a buffer leak possible here?

Reply:
No, it should be ok.